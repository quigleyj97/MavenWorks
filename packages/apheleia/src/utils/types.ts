export interface LayoutPartData {
    type: 'layout-part';
    title: string;
    showTitlebar: boolean;
    id: string;
}

export interface StackPanelData {
    type: 'stack-panel';
    direction: 'horizontal' | 'vertical';
    children: BaseRegionData[];
    stackSizes: string[];
}

export type BaseRegionData = LayoutPartData | StackPanelData;

export interface LayoutData {
    root: BaseRegionData;
}
