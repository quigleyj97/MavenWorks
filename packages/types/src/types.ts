import { GenericType } from "./generic";
import { Type } from "./type";

/**
 * Default built-in type definitions.
 *
 * @export
 * @class Types
 */
export class Types {
    /** The top type, aka "Any".
     *
     * This may represent any value, and any value may be assigned to it.
     *
     * @static
     * @memberof Types
     */
    public static Unknown = Type.CreateType("Unknown");

    /** A generic number type.
     *
     * @static
     * @memberof Types
     */
    public static Number = Type.CreateType("Number", Types.Unknown);
    /** A type for an integer.
     *
     * In JS, values of this type should be BigInts.
     *
     * @static
     * @memberof Types
     */
    public static Integer = Type.CreateType("Integer", Types.Number);
    /** A double-precision floating-point type.
     *
     * @static
     * @memberof Types
     */
    public static Float = Type.CreateType("Float", Types.Number);
    /**
     * A decimal type.
     *
     * This is useful for avoiding floating-point errors in decimal math, eg
     * financial pricing data. In JS, values of this type are Decimal.js
     * instances. cf. https://github.com/MikeMcl/decimal.js/
     *
     * @static
     * @memberof Types
     */
    public static Decimal = Type.CreateType("Decimal", Types.Number);
    /**
     * A string type.
     *
     * @static
     * @memberof Types
     */
    public static String = Type.CreateType("String", Types.Unknown);

    /**
     * An Object type, representing values that are not considered 'primitives'.
     *
     * @static
     * @memberof Types
     */
    public static Object = Type.CreateType("Object", Types.Unknown);

    /**
     * A Date type, representing a calendar date with no associated time.
     *
     * @static
     * @memberof Types
     */
    public static Date = Type.CreateType("Date", Types.Object);

    /**
     * A DateTime type, representing a combined calendar date and time-of-day.
     *
     * @static
     * @memberof Types
     */
    public static DateTime = Type.CreateType("DateTime", Types.Object);

    /**
     * A Time type, representing a time-of-day (but not a calendar date).
     *
     * @static
     * @memberof Types
     */
    public static Time = Type.CreateType("Time", Types.Object);

    /**
     * A generic factory to create Array types.
     *
     * @static
     * @memberof Types
     */
    public static Array = GenericType.CreateGenericFactory("Array", Types.Object, 1);

    /**
     * A generic factory to create Map types.
     *
     * @static
     * @memberof Types
     */
    public static Map = GenericType.CreateGenericFactory("Map", Types.Object, 2);
}