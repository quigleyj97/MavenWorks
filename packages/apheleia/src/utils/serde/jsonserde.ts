import { LayoutData } from "../types";
import { ILayoutSerde } from "./interface";

interface Data {
    version: {
        major: number,
        minor: number
    },
    layout: LayoutData;
};

export class JsonLayoutSerde implements ILayoutSerde {
    private static readonly VERSION_MAJOR = 0;
    private static readonly VERSION_MINOR = 1;

    public loadFromString(data: string): LayoutData {
        let root: Data;
        try {
            root = JSON.parse(data) as Data;
        } catch (err) {
            throw Error("Failed to deserialize JSON layout: " + err);
        }
        // for now just trust that it's layout data
        // TODO: implement a validation scheme like io-ts
        if (
            root.version.major > JsonLayoutSerde.VERSION_MAJOR ||
            root.version.minor > JsonLayoutSerde.VERSION_MINOR
        ) {
            throw Error("Data version exceeds this version");
        }
        return root.layout;
    }

    dumpToString(layout: LayoutData): string {
        return JSON.stringify({
            version: {
                major: JsonLayoutSerde.VERSION_MAJOR,
                minor: JsonLayoutSerde.VERSION_MINOR
            },
            layout: layout
        });
    }

}