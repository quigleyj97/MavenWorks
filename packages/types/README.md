# Types

Types, type checking, and type comparison mechanisms
