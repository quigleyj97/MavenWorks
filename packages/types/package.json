{
  "name": "@mavenomics/types",
  "version": "0.1.2",
  "description": "Types and Type Serialization",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "repository": "https://github.com/Mavenomics/MavenWorks",
  "author": "Mavenomics, Inc",
  "license": "GPL-3.0",
  "files": [
    "lib/"
  ],
  "scripts": {
    "build": "tsc --build",
    "build:ci": "echo pass",
    "clean": "rimraf ./lib *.tsbuildinfo",
    "lint": "tslint --project .",
    "test": "jest",
    "docs": "typedoc --tsconfig ./docs.tsconfig.json"
  },
  "devDependencies": {},
  "dependencies": {}
}
