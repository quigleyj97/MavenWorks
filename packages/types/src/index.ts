export { Types } from "./types";
export { Type } from "./type";
export { GenericType } from "./generic";
