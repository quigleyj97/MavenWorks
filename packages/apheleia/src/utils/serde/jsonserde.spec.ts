import * as fs from "fs";
import { LayoutData } from "../types";
import { JsonLayoutSerde } from "./jsonserde";

const TEST_FILE = fs.readFileSync("./testdata/layout.json", "utf8");

const TEST_LAYOUT: LayoutData = {
    root: {
        type: 'stack-panel',
        children: [{
            type: 'layout-part',
            showTitlebar: false,
            title: 'test-part',
            id: 'test-part'
        }],
        stackSizes: [
            "200px"
        ],
        direction: 'vertical'
    }
};

describe("JsonLayoutSerde tests", () => {
    let layoutSerde = new JsonLayoutSerde();

    test("loadFromString", () => {
        expect(layoutSerde.loadFromString(TEST_FILE)).toMatchSnapshot();
    });

    test("dumpToString", () => {
        expect(layoutSerde.dumpToString(TEST_LAYOUT)).toMatchSnapshot();
    });
});
