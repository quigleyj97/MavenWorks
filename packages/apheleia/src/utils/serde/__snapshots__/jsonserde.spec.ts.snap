// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonLayoutSerde tests dumpToString 1`] = `"{\\"version\\":{\\"major\\":0,\\"minor\\":1},\\"layout\\":{\\"root\\":{\\"type\\":\\"stack-panel\\",\\"children\\":[{\\"type\\":\\"layout-part\\",\\"showTitlebar\\":false,\\"title\\":\\"test-part\\",\\"id\\":\\"test-part\\"}],\\"stackSizes\\":[\\"200px\\"],\\"direction\\":\\"vertical\\"}}}"`;

exports[`JsonLayoutSerde tests loadFromString 1`] = `
Object {
  "root": Object {
    "children": Array [],
    "stackSizes": Array [],
    "type": "stack-panel",
  },
}
`;
